### DISCUSSION TITLE
Render Component on Condition

### DISCUSSION THREAD

**Discussion by codelover2k** [vercel/next.js#17388] [Help] [answered]
I am using getServerSideProps and there I check if the user is authenticated or not.
If the user is not authenticated I redirect and everything works so far, but the component, with a hook that fetches data from the database is still triggered.

Is there a way to tell getServerSideProps not avoid the initial render and just ignoring it?

**Comment by jamesmosier** [comment]
Your hook should not be being called server side. Where is your redirect code? In `getServerSideProps`?

```js
export function getServerSideProps(context) {
  const { res } = context;
  res.setHeader("location", "/my-other-page");
  res.statusCode = 302;
  res.end();
  return {props:{}}
}
```

---

If your hook is being call on the client side still then you should add some logic to it to prevent that. For example in `getServerSideProps` you could do: `return { props: { isLoggedIn: false } }` and then in your hook in your component you should check `if (props.isLoggedIn) { /* fetch data inside your hook */ }`

**Comment by codelover2k** [comment]
thanks for response James!

My code looks like yours for the redirection, I just do not use res.end() and the redirection works anyway.
If I use res.end(); I get this error:
`Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client
`
What I noticed is that with res.end() the component is not touched at all, what my aim is, but it looks like the res.end() error I get is a known bug.

P.S.:
I just figured out that you can wrap your hooks inside a condition in Next.js and it does not complain.
Not sure why

**Reply by jamesmosier** [reply]
Do you happen to have some code you can share which demonstrates the issue?

**Comment by codelover2k** [comment, marked as answer]
sure thing!

``` js
const Users = (props) => {
  const { loading, error, data } = useQuery(QUERY, {
    variables: { where: {} },
  });

  const { users = [] } = data;

  if (loading) {
    return <Loading />;
  }

  return (
    <>
      <Title>Users</Title>
      <Table data={users} />
    </>
  );
};


export const getServerSideProps = async (context) => {
 const { req, res } = context;
 const sessionUserExists = req && req.sessionUser && req.sessionUser.id;

 if (!sessionUserExists) {
  res.setHeader("location", "/");
  res.statusCode = 302;
  // with res.end() I get this error: 
  // Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client
  // res.end();
  return { props: {} };
 }

 const apolloClient = initializeApollo({ request: req });

 await apolloClient.query({
   query: QUERY,
   variables: { where: {} },
 });

 return {
   props: {
     initialApolloState: apolloClient.cache.extract(),
     sessionUser: req.sessionUser
   }
 };
};

export default Users;
```

I found a solution that works but it adds some noise:
I just use this code in my getServerSideProps:
```js
 if (!sessionUserExists) {
  res.setHeader("location", "/");
  res.statusCode = 302;
  return { props: { redirect: true } };
 }
```

and in my Users component I do this:
``` js
const Users = (props) => {

  if (props.redirect) {
    return null;
  }

...
```

**Comment by jamesmosier** [comment]
You could just check in your component `if (!props.sessionUser) { return null; }`. Then you don't have to set a `redirect` prop. If you don't get a `sessionUser` in your component then you know you need to redirect.

Or you could even just redirect client side using the same idea:

```js
import Router from 'next/router';

const Users = (props) => {
  useEffect(() => {
    if (!props.sessionUser) {
      Router.push('/');
    }
  }, []);

  if (!props.sessionUser) {
    return null;
  }

  return <div>your component</div>
};
```

**Comment by codelover2k** [comment]
thanks!

The client side solution is okay as long you have some small components, but I try to avoid that and do as much on server side as possible to give the user a greater UX 🙂

I asked previously this question and I hope you can help:
- is it normal that in Next.js you can put hooks inside conditions?

**Reply by jamesmosier** [reply]
No, that is not allowed in any React application. I'd have to see an example though to know for certain what might be going on there.

**Reply by codelover2k** [reply]
thanks for your help James, I used apollos query hook in a condition and it did not complain but I changed that because the `return null;`solution is cleaner , so I am happy for now 😁
